### ISSUE TITLE
Define a `resample` function that other stats packages can extend? 

### ISSUE THREAD

**Issue by kahaaga** [JuliaStats/StatsBase.jl#512]
In my [UncertainData.jl](https://github.com/kahaaga/UncertainData.jl) package, uncertainties are dealt with using a resampling approach. To have a common way of resampling across the package, I define a `resample` function, with different `resample` methods being used to resample uncertain values, or to resample datasets or subsets of datasets characterised by various types of uncertainties (e.g. theoretical distributions from Distributions.jl, values represented by kernel density estimates from empirical distributions, weighted distributions and more). 

It was suggested in [this thread](https://github.com/kahaaga/UncertainData.jl/issues/24#issuecomment-457037648) that it might be a good idea to have StatsBase define `resample`, then have UncertainData.jl and other packages in the statistics ecosystem extend it. This way, one could plugin various bootstrap methods (e.g. [Jackknife.jl](https://github.com/ararslan/Jackknife.jl) and [Bootstrap.jl](https://github.com/juliangehring/Bootstrap.jl)) across packages. 

I imagine it might lead to some issues if some other stats package also wants to define `resample`. Would it be a good idea to have a common interface for resampling methods in StatsBase.jl instead? If so, I could prepare a PR for it. I'm new to contributing to the community, so any thoughts are appreciated! 

### Proposed implementation

Just for reference, I imagine something something like this:

 In StatsBase.jl:

```julia
""" 
    AbstractResampling

Abstract type for resampling methods.
""" 
abstract type AbstractResampling end

"""
    AbstractSampling

Abstract type for the result of a resampling.
"""
abstract type AbstractSampling end

""" 
    resample

Resampling interface.
""" 
function resample end 
```

Then, the various stats packages can just extend `resample` for the actual resampling methods, `AbstractResampling` for defining resampling schemes (e.g. jackknifing), and `AbstractSampling` when defining types that hold the results of a resampling. 

For example, Bootstrap.jl might define 

```julia
import StatsBase: resample

""" 
    resample(f::Function, method::BootstrapSampling, data) -> BootstrapSample

Resample `data` using the provided bootstrap `method`, applying the point estimator 
`f` to each sample. 
""" 

function resample(f::Function, method::BootstrapSampling, data)
    bootstrap(f, data, method)
end
```

Jackknife.jl might define 

```julia 
import StatsBase: resample, AbstractResampling, AbstractSampling

""" 
Abstract type for jackknife resampling methods
"""
abstract type JackknifeResampling <: AbstractResampling
    
"""
A type indicating that leave-one-out resampling should be performed.
"""
struct LeaveOneOutResampling <: JackknifeResampling end

""" 
    JackknifeSampling

Composite type that holds the result of a jackknife resampling. 

## Fields 
- **`estimates`**: The `n-1` jackknifed estimates of `statistic` computed from `data`. 
- **`statistic`**: A point estimator of some statistic, for example `Statistics.mean`. 
- **`method`**: The jackknife resampling method, for example `LeaveOneOutResampling`.
- **`data`**: The data from which we compute estimate the statistic. 
""" 
struct JackknifeSampling{T, DT} <: AbstractSampling
    estimates::Vector{T}
    statistic::Function
    method::JackknifeResampling
    data::DT
end

""" 
    resample(f::Function, method::LeaveOneOutResampling, data)

Compute the jackknife estimate to a statistic by resampling `data` using 
leave-one-out sampling and applying the point estimator `f` to each 
of the samples.
""" 
function resample(f::Function, method::LeaveOneOutResampling, data)
    estims = Jackknife.leaveoneout(f, data)
    JackknifeSampling(estims, f, method, data)
end
```

In UncertainData.jl, one might for example define 

```julia
import StatsBase: resample
import Bootstrap: BasicSampling

""" 
    resample(f::Function, method::BasicSampling, uval::AbstractUncertainValue) -> ResampledStatistic 

Draw subsets/realizations of the uncertain value `uval` using the provided resampling `method` and 
apply the point estimator `f` to each realization.
""" 
function resample(f::Function, method::BasicSampling, uval::AbstractUncertainValue)
    estim = f(rand(uval, method.nrun))
end
```

This way, the user could do this kind of thing for resampling uncertain datasets using various bootstrap approaches.  For example:

```julia
# Define a dataset consisting of uncertain values defined by uniform distributions
uvals = [UncertainValue(Uniform, rand(-10:10), rand(20:30)) for i = 1:10]
udata = UncertainDataset(uvals)

# Compute the mean of each value in udata by using a basic bootstrap with n=1000
resample(mean, BasicSampling(1000), udata)

# Compute the mean of each value in udata by using leave-one-out resampling 
# to compute the mean on each value
resample(mean, LeaveOneOutResampling(), udata)
```

**Comment by nalimilan** [comment]
@ararslan @juliangehring What do you think about this?

**Comment by ararslan** [comment]
I like it.

**Comment by rofinn** [comment]
I might be inclined to suggest 2 signatures.

`resample(f, data, method)`
`resample(data, method)`

In the first case we'd return the appropriate sample type as you're already proposing, but in the second case we'd return an iterator of the raw samples drawn. This is mostly just because I often want to inspect the samples being drawn for debugging purposes (e.g., the maximum entropy code in Bootstrap.jl).

**Comment by juliangehring** [comment]
Sounds good to me.
Looking at the [julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/index.html#Write-functions-with-argument-ordering-similar-to-Julia-Base-1), would the signature have to be `resample(function, data, method)` instead of `resample(function, method, data)`? Not sure if `method` counts as "type" or "everything else" in this case.

**Comment by kahaaga** [comment]
Thanks for the feedback! 

I also like the idea of two separate signatures. In UncertainData.jl, I often use the drawn samples directly and don’t necessarily need to perform some computation on them, so the second signature would be useful too. 

Sounds like we agree that a common resample interface is a good idea, then? 

## Argument ordering

Thanks for bringing up the ordering and clearing that up, @juliangehring and @rofinn. I didn't pay too much attention to that in my initial suggestion, so I seem to have the order of the arguments mixed according to the style guide. It says function arguments first, then types, then input not being mutated, then everything else. I interpret that as follows:

### First signature (returns output of type `T <: AbstractSampling`)

In the first signature Rory suggests, `function` would be the summary statistic applied to the raw draws, which have been drawn by a resampling `method`. The `method` would be an instance of a type `T1 <: AbstractResampling`, which could be for example `LeaveOneOutResampling()` or `BasicBootstrap(n_draws)`. Lastly, `data` would be input not being mutated.  

This would return an instance of some type `T2 <: AbstractSampling` in accordance with the provided `method` (i.e. `JackKnifeSampling` or `BasicBootstrapSampling` for the above examples). Thus, `method` does not dictate the result type. I can imagine several cases, at least in UncertainData, where the types of the results could be dictated by the type of `data` itself, not by the resampling method. Also, different combinations of `data` and `method` could give samplings of different types as results. 

Therefore `method` is "everything else", and the signature should be `resample(function, data, method)`.

### Second signature (returns the raw draws)

In the second function signature, `method` is again an instance of type `T1 <: AbstractResampling`, and `data` is input not being mutated. 

This would return the raw samples, so `method` again does not dictate the return type. In this case, `data` is "input not being mutated" and `method` is "everything else", so the signature should therefore be `resample(data, method)`.

In other words, I didn't follow the style guide consistently in the initial proposal, and I agree with you both that the signature should be `resample(function, data, method)` and `resample(data, method)`.
